spring-integration-core/src/main/java/org/springframework/integration/dispatcher/PartitionedDispatcher.java

```diff
@@ -145,13 +145,15 @@ public void shutdown() {
 
 	@Override
 	public boolean dispatch(Message<?> message) {
-		prePopulatedPartitionsIfAny();
+		if (this.partitions.isEmpty()) {
```

Comment:
This won't work, thread 2 might proceed before thread 1 has populated all partition execs, and NPE on the dispatcher returned by the `get()`.

Reply:
How about now?
See the latest commit.